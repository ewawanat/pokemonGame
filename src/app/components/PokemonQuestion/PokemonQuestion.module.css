.container {
    text-align: center;
    margin-top: 20px;
}

.imageContainer {
    position: relative;
    width: 200px;
    height: 200px;
    /* Fixed height to prevent layout shifts */
    margin: 0 auto;
    overflow: hidden;
    /* Ensure the image doesn't overflow */
}

.image {
    object-fit: contain;
    /* Ensure the image scales to fit within the container without stretching */
}

.silhouette {
    filter: brightness(0);
    object-fit: contain;
    /* Ensure silhouette image also scales properly */
}

.optionsContainer {
    margin-top: 20px;
}

.optionButton {
    margin: 5px;
    padding: 10px 20px;
    background-color: #007bff;
    color: white;
    border: none;
    border-radius: 5px;
    cursor: pointer;
    transition: background-color 0.3s ease;
    font-size: 1rem;
}

.optionButton:hover:not(:disabled) {
    background-color: #0056b3;
}

.optionButton:disabled {
    background-color: #cccccc;
    cursor: not-allowed;
}