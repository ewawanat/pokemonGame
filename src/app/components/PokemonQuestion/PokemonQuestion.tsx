import React from 'react';
import Image from 'next/image';
import { Pokemon } from '../../models/Pokemon';

import styles from './PokemonQuestion.module.css';

type PokemonQuestionProps = {
    currentPokemon: Pokemon;
    options: string[];
    showAnswer: boolean;
    handleAnswer: (answer: string) => void;
};

const PokemonQuestion: React.FC<PokemonQuestionProps> = ({
    currentPokemon,
    options,
    showAnswer,
    handleAnswer,
}) => {
    return (
        <div className={styles.container}>
            <div className={styles.imageContainer}>
                <Image
                    src={showAnswer ? currentPokemon.image : currentPokemon.silhouette}
                    alt={showAnswer ? currentPokemon.name : 'silhouette of a pokemon'}
                    width={200}
                    height={200}
                    className={showAnswer ? styles.image : styles.silhouette}
                    priority
                />
            </div>
            <div className={styles.optionsContainer}>
                {options.map((option) => (
                    <button
                        key={option}
                        onClick={() => handleAnswer(option)}
                        className={styles.optionButton}
                        disabled={showAnswer}
                    >
                        {option}
                    </button>
                ))}
            </div>
        </div>
    );
};

export default PokemonQuestion;
