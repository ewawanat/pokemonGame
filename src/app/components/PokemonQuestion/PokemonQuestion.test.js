import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import PokemonQuestion from './PokemonQuestion';

// Extend Jest with accessibility matcher
expect.extend(toHaveNoViolations);

// tell Jest to replace the next/image module with the mock definition.
jest.mock('next/image', () => ({
    __esModule: true,
    default: ({ src, alt, width, height, className }) => (
        <img src={src} alt={alt} width={width} height={height} className={className} />
    ),
}));

describe('PokemonQuestion Component', () => {
    const mockHandleAnswer = jest.fn();
    const pokemon = {
        name: 'Pikachu',
        image: 'http://localhost/images/pikachu.png',
        silhouette: 'http://localhost/images/pikachu-silhouette.png',
    };

    const options = ['Pikachu', 'Bulbasaur', 'Charmander', 'Squirtle'];
    it('renders the silhouette when showAnswer is false', () => {
        render(
            <PokemonQuestion
                currentPokemon={pokemon}
                options={options}
                showAnswer={false}
                handleAnswer={mockHandleAnswer}
            />
        );
        const image = screen.getByAltText('silhouette of a pokemon');
        expect(image).toHaveProperty('src', pokemon.silhouette);
    });

    it('renders the actual image when showAnswer is true', () => {
        render(
            <PokemonQuestion
                currentPokemon={pokemon}
                options={options}
                showAnswer={true}
                handleAnswer={mockHandleAnswer}
            />
        );
        const image = screen.getByAltText('Pikachu');
        expect(image).toHaveProperty('src', pokemon.image);
    });

    it('renders all options as buttons', () => {
        render(
            <PokemonQuestion
                currentPokemon={pokemon}
                options={options}
                showAnswer={false}
                handleAnswer={mockHandleAnswer}
            />
        );
        const pikachuButton = screen.getByRole('button', { name: /pikachu/i });
        const bulbasaurButton = screen.getByRole('button', { name: /bulbasaur/i });

        const charmanderButton = screen.getByRole('button', { name: /charmander/i });

        const squirtleButton = screen.getByRole('button', { name: /squirtle/i });

        expect(pikachuButton).toBeInTheDocument();
        expect(bulbasaurButton).toBeInTheDocument();
        expect(charmanderButton).toBeInTheDocument();
        expect(squirtleButton).toBeInTheDocument();

    });

    it('calls handleAnswer with the selected option when a button is clicked', async () => {
        render(
            <PokemonQuestion
                currentPokemon={pokemon}
                options={options}
                showAnswer={false}
                handleAnswer={mockHandleAnswer}
            />
        );

        const button = screen.getByText('Pikachu');
        await userEvent.click(button);

        expect(mockHandleAnswer).toHaveBeenCalledWith('Pikachu');
    });

    it('disables the buttons when showAnswer is true', () => {
        render(
            <PokemonQuestion
                currentPokemon={pokemon}
                options={options}
                showAnswer={true}
                handleAnswer={mockHandleAnswer}
            />
        );

        const pikachuButton = screen.getByRole('button', { name: /pikachu/i });
        const bulbasaurButton = screen.getByRole('button', { name: /bulbasaur/i });
        const charmanderButton = screen.getByRole('button', { name: /charmander/i });
        const squirtleButton = screen.getByRole('button', { name: /squirtle/i });

        expect(pikachuButton).toBeDisabled();
        expect(bulbasaurButton).toBeDisabled();
        expect(charmanderButton).toBeDisabled();
        expect(squirtleButton).toBeDisabled();
    });

    it('does not disable the buttons when showAnswer is false', () => {
        render(
            <PokemonQuestion
                currentPokemon={pokemon}
                options={options}
                showAnswer={false}
                handleAnswer={mockHandleAnswer}
            />
        );


        const pikachuButton = screen.getByRole('button', { name: /pikachu/i });
        const bulbasaurButton = screen.getByRole('button', { name: /bulbasaur/i });
        const charmanderButton = screen.getByRole('button', { name: /charmander/i });
        const squirtleButton = screen.getByRole('button', { name: /squirtle/i });

        expect(pikachuButton).not.toBeDisabled();
        expect(bulbasaurButton).not.toBeDisabled();
        expect(charmanderButton).not.toBeDisabled();
        expect(squirtleButton).not.toBeDisabled();
    });
    it('has no accessibility violations when showAnswer is false', async () => {
        const { container } = render(
            <PokemonQuestion
                currentPokemon={pokemon}
                options={options}
                showAnswer={false}
                handleAnswer={mockHandleAnswer}
            />
        );

        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });

    it('has no accessibility violations when showAnswer is true', async () => {
        const { container } = render(
            <PokemonQuestion
                currentPokemon={pokemon}
                options={options}
                showAnswer={true}
                handleAnswer={mockHandleAnswer}
            />
        );

        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });
    it('matches the snapshot', () => {

        const { asFragment } = render(
            <PokemonQuestion
                currentPokemon={pokemon}
                options={options}
                showAnswer={false}
                handleAnswer={mockHandleAnswer}
            />
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
