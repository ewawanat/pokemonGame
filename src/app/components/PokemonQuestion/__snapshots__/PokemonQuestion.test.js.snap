// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonQuestion Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="imageContainer"
    >
      <img
        alt="silhouette of a pokemon"
        class="silhouette"
        height="200"
        src="http://localhost/images/pikachu-silhouette.png"
        width="200"
      />
    </div>
    <div
      class="optionsContainer"
    >
      <button
        class="optionButton"
      >
        Pikachu
      </button>
      <button
        class="optionButton"
      >
        Bulbasaur
      </button>
      <button
        class="optionButton"
      >
        Charmander
      </button>
      <button
        class="optionButton"
      >
        Squirtle
      </button>
    </div>
  </div>
</DocumentFragment>
`;
